### DISCUSSION TITLE
SAFE_HEAP delay the linking process

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#16944] [Q&A]
when I use in SAFE_HEAP it delay the linking process (is take about 2m!) 
is there a way to decrease that time?
here is the linker flags to emcc:
`-s SAFE_HEAP=2    -s OFFSCREEN_FRAMEBUFFER=1 -s OFFSCREENCANVAS_SUPPORT=1   -s WASM_BIGINT  -g --preload-file com.skyline.terraexplorer --bind -s USE_PTHREADS -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_RUNTIME_METHODS=ccall,cwrap -s EXPORTED_FUNCTIONS=_openFly,_main,_MobileOnSurfaceChanged,_OnLButtonDown,_OnLButtonUp,_OnMouseMove -s PTHREAD_POOL_SIZE=16 `

**Comment by kripken** [comment]
if `SAFE_HEAP` is the only reason to process the wasm after linking, which I think is the case in your command, then it will add all the overhead of reading and writing the wasm, plus the safe-heap transform. However, those should be fast linear operations, so this is still odd. Running `EMCC_DEBUG=1` in the env will give some timing info which might be helpful (like maybe the time is in the JS side and not wasm).

Also, running a system profiler on `wasm-opt` might find something interesting. You can also try `-sBINARYEN_EXTRA_PASSES=no-validation`.

**Comment by aharondavid** [comment]
BINARYEN_EXTRA_PASSES doesn't help.
I will try the others.
Thanks a lot.
